File: src/main/java/calculator/CustomDelimiterTextExtractor.java
package calculator;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class CustomDelimiterTextExtractor implements TextExtractor {

  private final static String REGEX_PATTERN = "//(.)\n(.*)";
  private final static Pattern PATTERN = Pattern.compile(REGEX_PATTERN);

  @Override
  public boolean isSupport(String text) {
    return PATTERN.matcher(text)
        .find();
  }

  @Override
  public String[] extract(String text) {
    Matcher matcher = PATTERN.matcher(text);
    if (matcher.find()) {
      String customDelimiter = matcher.group(1);
      return matcher.group(2).split(customDelimiter);
    }
    throw new TextExtractionException("본문 추출에 실패했습니다.");
  }
}

File: src/main/java/calculator/extractor/CustomDelimiterTextExtractor.kt
package calculator.extractor

import calculator.Number
import calculator.Numbers
import calculator.exception.TextDelimiterExtractorException
import java.util.regex.Pattern

class CustomDelimiterTextExtractor : TextDelimiterExtractor {

    override fun isSupport(text: String?): Boolean {
        return text?.let { TWO_SLASHES_AND_NEWLINE_STRING_BETWEEN_DELIMITER_PATTERN.matcher(it).find() } ?: false
    }

    override fun extract(text: String): Numbers {
        val matcher = TWO_SLASHES_AND_NEWLINE_STRING_BETWEEN_DELIMITER_PATTERN.matcher(text)

        return if (matcher.find()) {
            val customDelimiter = matcher.group(CUSTOM_DELIMITER_FIND_INDEX)

            Numbers(
                matcher.group(CUSTOM_DELIMITER_VALUE_INDEX)
                    .split(customDelimiter)
                    .map { Number(it) }
            )
        } else {
            throw TextDelimiterExtractorException("Unsupported text format: $text")
        }
    }

    companion object {
        private val TWO_SLASHES_AND_NEWLINE_STRING_BETWEEN_DELIMITER_PATTERN = Pattern.compile("//(.)\n(.*)")
        private const val CUSTOM_DELIMITER_FIND_INDEX = 1
        private const val CUSTOM_DELIMITER_VALUE_INDEX = 2
    }
}
